### PR TITLE
Run compute-sanitizer in nightly build

### PR DESCRIPTION
**Pull request by karthikeyann**
Addresses part of https://github.com/rapidsai/cudf/issues/904
- This PR enables run of `compute-sanitizer --tool memcheck` on libcudf unit tests when env `COMPUTE_SANITIZER_ENABLE=true`
  This env `COMPUTE_SANITIZER_ENABLE` will be enabled only in nightly builds of cudf. (To be Enabled in PR https://github.com/rapidsai/gpuci-scripts/pull/675)
- This PR also adds script to parse compute-sanitizer log to junit xml file which can be processed by Jenkins.
  Reports only failures. If no errors, no tests are reported under memcheck results.

Note: Only `memcheck` is enabled now.  when required, other checks of compute-sanitizer could be enabled later.

### CODE REVIEW

**Comment by davidwendt** [comment]
Seems like this should not be going into `21.12` since we are well into burndown.

**Comment by karthikeyann** [comment]
rerun tests

**Comment by karthikeyann** [comment]
Preferably merge to 21.12 so that memchecks could run in nightly builds for previous branches as well.
Not enabled in gpuci nightly build yet. So, it's it won't be enabled in CI scripts.

**Comment by davidwendt** [comment]
> Preferably merge to 21.12 so that memchecks could run in nightly builds for previous branches as well. Not enabled in gpuci nightly build yet. So, it's not be enabled in PR CI runs.

The concern is even the faint possibly of breaking the build at a time when the build is critical (and build team over engaged) in the release cycle.

File: ci/gpu/build.sh [hunk 1]
```diff
@@ -176,6 +176,19 @@ else
         ${gt} --gtest_output=xml:"$WORKSPACE/test-results/"
     done
 
+    if [[ "$COMPUTE_SANITIZER_ENABLE" == "1" ]]; then
+        gpuci_logger "Memcheck on GoogleTests with rmm_mode=cuda"
+        export GTEST_CUDF_RMM_MODE=cuda
+        COMPUTE_SANITIZER_CMD="compute-sanitizer --tool memcheck --log-file" #  $WORKSPACE/test-results/cudamemcheck-%q{test_name}.log
```
Comment (mythrocks) [inline ci/gpu/build.sh:182]: Did we intend to remove that `$WORKSPACE` bit?
```suggestion
        COMPUTE_SANITIZER_CMD="compute-sanitizer --tool memcheck --log-file"
```
Or is it to illustrate how `$COMPUTE_SANITIZER_CMD` is completed when used?

Comment (mythrocks) [reply ci/gpu/build.sh:182]: Also, update Copyright year?

File: cpp/scripts/compute-sanitizer-to-junit-xml.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+def parse_args():
+    argparser = argparse.ArgumentParser(
+        "Translate and collates compute-sanitizer logs of googletest/pytest unit-tests to junit xml"
+    )
+    argparser.add_argument(
+        "-out",
+        type=str,
+        default=None,
+        required="-glob" in sys.argv,
+        help="Output compute-sanitizer juni xml file path",
```
Comment (mythrocks) [inline cpp/scripts/compute-sanitizer-to-junit-xml.py:36]:
```suggestion
        help="Output compute-sanitizer junit xml file path",
```

File: cpp/scripts/compute-sanitizer-to-junit-xml.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                        print("Unknown line: " + line)
+                elif line.startswith("[ RUN      ] "):  # gtest testcase
+                    testcase_name = line.split("[ RUN      ] ")[1].strip()
+                elif line.startswith("[       OK ] "):
+                    testcase_name = ""
+                elif line.startswith("[  FAILED  ] "):
+                    testcase_name = ""
+                else:
+                    pass
+                    # raise Exception('unexpected line in compute-sanitizer log: '+line)
```
Comment (mythrocks) [inline cpp/scripts/compute-sanitizer-to-junit-xml.py:130]: Hmm. If none of the above, should the tool pass? 
I think I can see why we'd leave this comment in. Might be useful to uncomment for debugging.

**Review by mythrocks — APPROVED** [review]
A couple of nitpicks.

Bigger picture: I was wondering why we're converting to the `junit` format.

**Comment by karthikeyann** [comment]
> Bigger picture: I was wondering why we're converting to the `junit` format.

so that Jenkins can display it similar to how gtest results are shown. Jenkins junit plugin expects the results in XML.

File: cpp/scripts/compute-sanitizer-to-junit-xml.py (hunk 3 above)
Comment (karthikeyann) [reply cpp/scripts/compute-sanitizer-to-junit-xml.py:130]: Yes. The log may contain stdout from googletest or pytest which tool will ignore (pytest is not memchecked yet, but will also be too large to add, probably not nightly, may be weekly).
This is added for debug for analyzing only memcheck log. (using `--log-file path`)

File: ci/gpu/build.sh (hunk 1 above)
Comment (karthikeyann) [reply ci/gpu/build.sh]: updated.

**Comment by karthikeyann** [comment]
Thank you for the reviews @mythrocks  and @davidwendt

File: ci/gpu/build.sh [hunk 4]
```diff
@@ -176,6 +176,25 @@ else
         ${gt} --gtest_output=xml:"$WORKSPACE/test-results/"
     done
 
+    ################################################################################
+    # MEMCHECK - Run compute-sanitizer on GoogleTest (only in nightly builds)
+    ################################################################################
+    if [[ "$BUILD_MODE" == "branch" && "$BUILD_TYPE" == "gpu" ]]; then
+        if [[ "$COMPUTE_SANITIZER_ENABLE" == "1" ]]; then
```
Comment (karthikeyann) [inline ci/gpu/build.sh:183]:
```suggestion
        if [[ "$COMPUTE_SANITIZER_ENABLE" == "true" ]]; then
```

**Comment by karthikeyann** [comment]
Tested these environment variables in local docker image after simplifying the PR https://github.com/rapidsai/gpuci-scripts/pull/675

**Comment by karthikeyann** [comment]
rerun tests

**Comment by karthikeyann** [comment]
rerun tests

**Comment by karthikeyann** [comment]
rerun tests

**Comment by karthikeyann** [comment]
rerun tests

**Comment by karthikeyann** [comment]
rerun tests

**Comment by karthikeyann** [comment]
memcheck failures reported example.
https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-gpu-test/CUDA=11.5,GPU_LABEL=cuda115,LINUX_VER=centos7,PYTHON=3.8/5082/

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

**Comment by karthikeyann** [comment]
rerun tests

**Comment by karthikeyann** [comment]
@gpucibot merge

File: ci/gpu/build.sh [hunk 5]
```diff
@@ -176,6 +176,28 @@ else
...
+            export GTEST_CUDF_RMM_MODE=cuda
+            COMPUTE_SANITIZER_CMD="compute-sanitizer --tool memcheck"
+            mkdir -p "$WORKSPACE/test-results/"
+            for gt in gtests/*; do
+                test_name=$(basename ${gt})
+                if [[ "$test_name" == "ERROR_TEST" ]]; then
+                  continue
+                fi
+                echo "Running GoogleTest $test_name"
+                ${COMPUTE_SANITIZER_CMD} ${gt} | tee "$WORKSPACE/test-results/${test_name}.cs.log"
```
Comment (davidwendt) [inline ci/gpu/build.sh:194]: Could this have been the following:
```
 ${COMPUTE_SANITIZER_CMD} ${gt} --rmm_mode=cuda | tee "$WORKSPACE/test-results/${test_name}.cs.log"
```
And then you would not need to set and unset the `GTEST_CUDF_RMM_MODE` environment variable?

Comment (jrhemstad) [reply ci/gpu/build.sh:194]: We should use the cuda async mode for RMM. That will help the tests run faster by using the CUDA pool but still have memcheck support.

**Pull request merged**